src/common/log_file.c

```diff
@@ -0,0 +1,167 @@
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "common/log.h"
+
+/* emit inline symbol */
+extern bool ardop_logfile_need_rollover(
+	const time_t current_file_tm,
+	const time_t now
+);
+
+void ardop_logfile_init(
+	ArdopLogFile* logfile,
+	const char* logdir,
+	const char* stem,
+	const char* extension,
+	const uint16_t port
+) {
+	memset(logfile, 0, sizeof(*logfile));
+	logfile->logdir = logdir;
+	logfile->stem = stem;
+	logfile->extension = extension;
+	logfile->port = port;
+	logfile->opened_at = INT32_MIN;
+}
+
+bool ardop_logfile_write(
+	ArdopLogFile* logfile,
+	const void* msg,
+	const size_t msglen)
+{
+	struct timespec tp = { 0, 0 };
+	clock_gettime(CLOCK_REALTIME, &tp);
+
+	FILE* outfile = ardop_logfile_handle(logfile, tp.tv_sec);
+	if (!!outfile) {
+		size_t nwritten = fwrite(msg, msglen, 1, outfile);
+		fflush(outfile);
+		return (nwritten == msglen);
+	}
+	else
+	{
+		return false;
+	}
+}
+
+bool ardop_logfile_write_norotate(
+	ArdopLogFile* logfile,
+	const void* msg,
+	const size_t msglen)
+{
+	if (!! logfile->handle) {
+		size_t nwritten = fwrite(msg, msglen, 1, logfile->handle);
+		fflush(logfile->handle);
+		return (nwritten == msglen);
+	}
+	else
+	{
+		return false;
+	}
+}
+
+int ardop_logfile_close(ArdopLogFile* logfile) {
+	logfile->opened_at = INT32_MIN;
+
+	if (!! logfile->handle) {
+		int rv = fclose(logfile->handle);
+		logfile->handle = 0;
+		return rv;
+	}
+	else
+	{
+		return 0;
+	}
+}
+
+FILE* ardop_logfile_handle(
+	ArdopLogFile* logfile,
+	const time_t now)
+{
+	// no port number → no log file
+	if (!logfile->port) {
+		return NULL;
+	}
+
+	// previous failed open attempt → no log file
+	if (!logfile->handle && logfile->opened_at > INT32_MIN) {
+		return NULL;
+	}
+
+	// open log if there is none or it is time to rollover
+	if (!logfile->handle || ardop_logfile_need_rollover(logfile->opened_at, now)) {
+		logfile->opened_at = now;
+
+		char log_file_name[1024] = "";
+		if (ardop_logfile_calc_filename(
+			logfile,
+			now,
+			log_file_name,
+			sizeof(log_file_name)))
+		{
+			if (! logfile->handle) {
+				logfile->handle = fopen(log_file_name, "a");
+			} else {
+				// close the old log file and open the new one
+				// the glibc implementation of this is atomic with dup3(2),
+				// but the standard does not guarantee this
+				logfile->handle = freopen(log_file_name, "a", logfile->handle);
+			}
+
+			if (!logfile->handle) {
+				ZF_LOGE("Unable to open log file: \"%s\": %s", log_file_name, strerror(errno));
+			}
+		}
+		else
+		{
+			logfile->opened_at = INT32_MIN + 1;
+			ZF_LOGE("Unable to determine log file name");
+		}
+	}
+
+	return logfile->handle;
+}
+
+ARDOP_MUSTUSE bool ardop_logfile_calc_filename(
+	ArdopLogFile* logfile,
+	const time_t now,
+	char* out,
+	const size_t outsize
+) {
+	// convert time_t to calendar time
+	struct tm cal;
+	localtime_r(&now, &cal);
+
+	// format time
+	char cal_fmt[32];
+	strftime(cal_fmt, sizeof(cal_fmt), "%Y%m%d_%Z", &cal);
```

Comment:
We should prefer `gmtime_r()` over `gmtime()` since it is MT-safe (multithread safe). Granted, we're not writing a thread-safe logger, but I see little reason to add gotchas for future me.

I have removed the time zone from log file names.

Reply:
Personally, I think this is a pointless choice, but I don't think it hurts anything.  So, I'll accept it.  gmtime() is being used elsewhere [ALSASound.c:StartTxWav())] and ardopcf is unlikely to ever be a multithreaded program.  A multithreaded version would allow complete restructuring of the RX and TX audio signal processing paths.  Thus, to properly take advantage of multithreading would be a major rewrite, at which point it probably makes sense to build a completely new program.